velox/exec/RowsStreamingWindowBuild.cpp

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/exec/RowsStreamingWindowBuild.h"
+#include "velox/exec/RowsStreamingWindowPartition.h"
+
+namespace facebook::velox::exec {
+
+RowsStreamingWindowBuild::RowsStreamingWindowBuild(
+    const std::shared_ptr<const core::WindowNode>& windowNode,
+    velox::memory::MemoryPool* pool,
+    const common::SpillConfig* spillConfig,
+    tsan_atomic<bool>* nonReclaimableSection)
+    : WindowBuild(windowNode, pool, spillConfig, nonReclaimableSection) {}
+
+void RowsStreamingWindowBuild::buildNextInputOrPartition(bool isFinished) {
+  if (windowPartitions_.size() <= inputCurrentPartition_) {
+    windowPartitions_.push_back(std::make_shared<RowsStreamingWindowPartition>(
+        data_.get(),
+        folly::Range<char**>(nullptr, nullptr),
+        inputColumns_,
+        sortKeyInfo_));
+  }
+
+  windowPartitions_[inputCurrentPartition_]->addNewRows(inputRows_);
+
+  if (isFinished) {
+    windowPartitions_[inputCurrentPartition_]->setInputRowsFinished();
+    inputCurrentPartition_++;
+  }
+
+  inputRows_.clear();
+}
+
+void RowsStreamingWindowBuild::addInput(RowVectorPtr input) {
+  for (auto i = 0; i < inputChannels_.size(); ++i) {
+    decodedInputVectors_[i].decode(*input->childAt(inputChannels_[i]));
+  }
+
+  for (auto row = 0; row < input->size(); ++row) {
+    char* newRow = data_->newRow();
+
+    for (auto col = 0; col < input->childrenSize(); ++col) {
+      data_->store(decodedInputVectors_[col], row, newRow, col);
+    }
+
+    if (previousRow_ != nullptr &&
+        compareRowsWithKeys(previousRow_, newRow, partitionKeyInfo_)) {
+      buildNextInputOrPartition(true);
+    }
+
+    // Wait for the peers to be ready in single partition; these peers are the
```

Comment:
The condition inputRows_.size() >= numRowsPerOutput is hard to understand. What does it handle ?

Reply:
@aditi-pandit 
In our internal testing, we found that without this restriction, queries q57 and q47 in TPC-DS would result in an OOM condition. Upon observation, the values in the ORDER BY clause are all distinct, meaning that each partial window partition contains only one row of data. This leads to the getOutput method in window.cpp processing only one row of data for each window partition computation. We speculate that this occurs because each computation requires the allocation of a frame buffer and peer buffer, and the size of these buffers is determined based on the value of numRowsPerOutput_. This results in holding too many buffers, while in reality, only the buffer for a single row of data is used, and the buffers allocated for numRowsPerOutput_ are not utilized. Therefore, we have added this restriction.

Reply:
Can the check fix the issue? I still saw OOM if each partition is only one row